sklearn/utils/_metadata_requests.py

```diff
@@ -0,0 +1,1157 @@
+                    k,
+                    inspect.Parameter.KEYWORD_ONLY,
+                    default=UNCHANGED,
+                    annotation=Optional[Union[RequestType, str]],
+                )
+                for k in self.keys
+            ]
+        )
+        func.__signature__ = inspect.Signature(
+            params,
+            return_annotation=owner,
+        )
+        doc = REQUESTER_DOC.format(method=self.name)
+        for metadata in self.keys:
+            doc += REQUESTER_DOC_PARAM.format(metadata=metadata, method=self.name)
+        doc += REQUESTER_DOC_RETURN
+        func.__doc__ = doc
+        return func
+
+
+class _MetadataRequester:
+    """Mixin class for adding metadata request functionality.
+
+    .. versionadded:: 1.1
+    """
+
+    def __init_subclass__(cls, **kwargs):
+        """Set the ``set_{method}_request`` methods.
+
+        This uses PEP-487 [1]_ to set the ``set_{method}_request`` methods. It
+        looks for the information available in the set default values which are
+        set using ``__metadata_request__*`` class attributes.
+
+        References
+        ----------
+        .. [1] https://www.python.org/dev/peps/pep-0487
+        """
+        try:
+            requests = cls._get_default_requests()
+        except Exception:
+            # if there are any issues in the default values, it will be raised
+            # when ``get_metadata_routing`` is called. Here we are going to
+            # ignore all the issues such as bad defaults etc.
+            super().__init_subclass__(**kwargs)
+            return
+
+        for method in METHODS:
+            mmr = getattr(requests, method)
+            # set ``set_{method}_request``` methods
+            if not len(mmr.requests):
+                continue
+            setattr(
+                cls,
+                f"set_{method}_request",
+                RequestMethod(method, sorted(mmr.requests.keys())),
+            )
+        super().__init_subclass__(**kwargs)
+
+    @classmethod
+    def _get_default_requests(cls):
+        """Collect default request values.
+
+        This method combines the information present in ``metadata_request__*``
+        class attributes.
+        """
+
+        requests = MetadataRequest(owner=cls.__name__)
+
+        # need to go through the MRO since this is a class attribute and
+        # ``vars`` doesn't report the parent class attributes. We go through
+        # the reverse of the MRO since cls is the first in the tuple and object
+        # is the last.
+        defaults = dict()
+        for klass in reversed(inspect.getmro(cls)):
+            klass_defaults = {
+                attr: value
+                for attr, value in vars(klass).items()
+                if "__metadata_request__" in attr
+            }
+            defaults.update(klass_defaults)
+        defaults = dict(sorted(defaults.items()))
+
+        # First take all arguments from the method signatures and have them as
```

Comment:
Maybe this chunk (or the loop's contents) should be pulled into a helper, e.g. `_build_request_from_signature`.